GenderPayGap.WebUI/Controllers/Admin/AdminDatabaseIntegrityChecksController.cs

```diff
@@ -200,5 +201,192 @@ public IActionResult PrivateSectorOrganisationsWithAPublicSectorType()
             return PartialView("PrivateSectorOrganisationsWithAPublicSectorType", privateSectorOrganisationsWithAPublicSectorType);
         }
 
+        [HttpGet("database-integrity-checks/returns-with-figures-with-more-than-one-decimal-place")]
+        public IActionResult ReturnsWithFiguresWithMoreThanOneDecimalPlace()
+        {
+            List<Return> invalidReturns =
+                dataRepository.GetAll<Return>()
+                    .Where(r => r.Status == ReturnStatuses.Submitted)
+                    .AsEnumerable()
+                    .Where(
+                        r => HasMoreThanOneDecimalPlace(r.FemaleLowerPayBand)
+                             || HasMoreThanOneDecimalPlace(r.MaleLowerPayBand)
+                             || HasMoreThanOneDecimalPlace(r.FemaleMiddlePayBand)
+                             || HasMoreThanOneDecimalPlace(r.MaleMiddlePayBand)
+                             || HasMoreThanOneDecimalPlace(r.FemaleUpperPayBand)
+                             || HasMoreThanOneDecimalPlace(r.MaleUpperPayBand)
+                             || HasMoreThanOneDecimalPlace(r.FemaleUpperQuartilePayBand)
+                             || HasMoreThanOneDecimalPlace(r.MaleUpperQuartilePayBand)
+                             || HasMoreThanOneDecimalPlace(r.FemaleMedianBonusPayPercent)
+                             || HasMoreThanOneDecimalPlace(r.MaleMedianBonusPayPercent)
+                             || HasMoreThanOneDecimalPlace(r.DiffMeanBonusPercent)
+                             || HasMoreThanOneDecimalPlace(r.DiffMedianBonusPercent)
+                             || HasMoreThanOneDecimalPlace(r.DiffMedianHourlyPercent)
+                             || HasMoreThanOneDecimalPlace(r.DiffMeanHourlyPayPercent))
+                    .ToList();
+
+            return PartialView("ReturnsWithFiguresWithMoreThanOneDecimalPlace", invalidReturns);
+        }
+
+        [HttpGet("database-integrity-checks/returns-with-quarters-figures-sum-different-than-100")]
+        public IActionResult ReturnsWithQuartersFiguresSumDifferentThan100()
+        {
+            List<Return> invalidReturns =
+                dataRepository.GetAll<Return>()
+                    .Where(r => r.Status == ReturnStatuses.Submitted)
+                    .AsEnumerable()
```

Comment:
I'm fairly sure (maybe worth checking!!) that this `AsEnumerable` call will download all the submitted returns into memory and further parse them there. This obviously poses issues if you've got an absolutely massive database (and is I assume why the other tests in this file except one don't use it, so Entity Framework automatically translates it the LINQ into SQL).

I think you ought to be able to some of these new tests entirely without `AsEnumerable` (see eg [here](https://stackoverflow.com/questions/25489331/get-the-sum-of-multiple-columns) for summing two fields, it's a bit like doing `SELECT (FemaleUpperPayBand + MaleUpperPayBand) AS sum WHERE sum != 100`. wdyt?

(It's a shame because I really like this style, it's really slick, I'm just interested if we can squeeze some extra performance out of it)

Reply:
Good point! I changed the queries so that they can be translated to SQL. There were 2 that cannot be translated so I kept them as they were (figure has more than 1 decimal place and company link is invalid)